testlapack/dlantr.go

```diff
@@ -0,0 +1,77 @@
+package testlapack
+
+import (
+	"math"
+	"math/rand"
+	"testing"
+
+	"github.com/gonum/blas"
+	"github.com/gonum/lapack"
+)
+
+type Dlantrer interface {
+	Dlanger
+	Dlantr(norm lapack.MatrixNorm, uplo blas.Uplo, diag blas.Diag, m, n int, a []float64, lda int, work []float64) float64
+}
+
+func DlantrTest(t *testing.T, impl Dlantrer) {
+	for _, norm := range []lapack.MatrixNorm{lapack.MaxAbs, lapack.MaxColumnSum, lapack.MaxRowSum, lapack.NormFrob} {
+		for _, diag := range []blas.Diag{blas.NonUnit, blas.Unit} {
+			for _, uplo := range []blas.Uplo{blas.Lower, blas.Upper} {
+				for _, test := range []struct {
+					m, n, lda int
+				}{
+					{3, 3, 0},
+					{3, 5, 0},
+					{10, 5, 0},
+
+					{5, 5, 11},
+					{5, 10, 11},
+					{10, 5, 11},
+				} {
+					// Do a couple of random trials since the values change.
+					for i := 0; i < 1; i++ {
+						m := test.m
+						n := test.n
+						lda := test.lda
+						if lda == 0 {
+							lda = n
+						}
+						a := make([]float64, m*lda)
+						for i := range a {
+							a[i] = rand.NormFloat64()
+						}
+						aDense := make([]float64, len(a))
+						if uplo == blas.Lower {
+							for i := 0; i < m; i++ {
+								for j := 0; j <= min(i, n-1); j++ {
+									aDense[i*lda+j] = a[i*lda+j]
+								}
+							}
+						} else {
+							for i := 0; i < m; i++ {
+								for j := i; j < n; j++ {
+									aDense[i*lda+j] = a[i*lda+j]
+								}
+							}
+						}
+						if diag == blas.Unit {
+							for i := 0; i < min(m, n); i++ {
+								aDense[i*lda+i] = 1
+							}
+						}
+						work := make([]float64, n)
+						for i := range work {
+							work[i] = rand.Float64()
+						}
+						nrm := impl.Dlantr(norm, uplo, diag, m, n, a, lda, work)
```

Comment:
Please use got and want here - when debugging and the world is uncertain, it makes life easier.